dht.go

```diff
@@ -133,33 +149,86 @@ func NewDHTClient(ctx context.Context, h host.Host, dstore ds.Batching) *IpfsDHT
 
 func makeDHT(ctx context.Context, h host.Host, dstore ds.Batching, protocols []protocol.ID) *IpfsDHT {
 	rt := kb.NewRoutingTable(KValue, kb.ConvertPeerID(h.ID()), time.Minute, h.Peerstore())
+	rtRecoveryChan := make(chan *rtRecoveryReq)
 
 	cmgr := h.ConnManager()
+
 	rt.PeerAdded = func(p peer.ID) {
 		cmgr.TagPeer(p, "kbucket", 5)
 	}
+
 	rt.PeerRemoved = func(p peer.ID) {
 		cmgr.UntagPeer(p, "kbucket")
+		go func(rtRecoveryChan chan *rtRecoveryReq) {
```

Comment:
There's a simpler solution:

First, add a channel (protected with a lock) to signal when bootstrapping is done. Tasks that need to wait for peers in the routing table can wait for this channel to be closed.

On peer add:

1. If the recovery channel is closed, return.
2. Otherwise, take the lock.
3. Check 1 again.
4. Close the channel (we've recovered).
5. Trigger a bootstrap run.

On peer remove:

1. Check if the channel is closed. If not, return as we're currently bootstrapping.
2. Check to see if the table is empty. If not, return as we have peers.
3. Take the lock, defer a release of the lock.
4. Re-run steps 1-2 with the lock held.
5. Replace the channel with a new _open_ channel.
6. Launch a recovery goroutine.

In the recovery goroutine, keep recovering, sleeping, recovering, etc. until the recovery channel is closed.

Note: this is just _one_ approach. The key parts are:

1. We launch fewer goroutines.
2. We stop the recovery process when we've _actually_ recovered (i.e., the routing table has entries).
3. We do very little work unless we actually _have_ work to do.

My concerns with the current approach are:

1. We launch a goroutine every time we drop a peer.
2. We consider the recovery process "done" when it finishes instead of when the routing table gains a peer.

Reply:
@Stebalien Have linked this comment in #387. This particular task will now be done as part of that issue.